adcc/adc_pp/matrix.py

```diff
@@ -39,15 +38,53 @@
 #      really so much our focus.
 
 
-#
-# Dispatch routine
-#
-"""
-`apply` is a function mapping an AmplitudeVector to the contribution of this
-block to the result of applying the ADC matrix. `diagonal` is an `AmplitudeVector`
-containing the expression to the diagonal of the ADC matrix from this block.
-"""
-AdcBlock = namedtuple("AdcBlock", ["apply", "diagonal"])
```

Comment:
Whenever I add 60 lines of code I ask myself: Is it worth it. And here I am not so sure. The only thing you really need is ` return sum(app(ampl) for app in self._applies)` as the diagonal summation is done by the AdcMatrix class later anyway. And for the sum you could also just generate a `lambda ampl: sum(app(ampl) for app in self._applies)` and use that instead of the already existing apply function inside the AdcMatrix for the patched block. Given that this version is not much slower (which we should test) I would prefer that, because that keeps also the "block logic" in the AdcMatrix where it belongs and shortens this back to 5 lines instead of 65.

Reply:
I'm not sure I understood everything...
Do you want to revert to `namedtuple`, if yes, where would `self._applies` come from?

Reply:
Nowhere. Instead of "adding the block" you form a lambda inside the AdcMatrix class or so that does the addition between the applies.

Reply:
I'm still a bit uncertain...
I tried the following now
```Python
for sp in other.blocks:
    original_apply = self.blocks_ph[sp].apply
    def patched_apply(ampl):
        return sum(
            app(ampl)
            for app in (original_apply, other.blocks[sp].apply)
        )
    self.blocks_ph[sp].apply = patched_apply
```
if that's what you meant... Problem is one cannot re-set the attributes of namedtuple.

Reply:
Ah crap ... I forgot about that. But then change the named tuple to a simple class with just two attributes?

Reply:
or maybe avoid storing the named tuple in the AdcMatrix and intstead only store the apply functions (and the diagonal is munched into the precomputed thing anyway)?

Reply:
Done.